Update quickstart tutorials and acknowledgements (#307)
learning-at-home/hivemind: README.md

[![PyPI version](https://img.shields.io/pypi/v/hivemind.svg)](https://pypi.org/project/hivemind/)
[![Discord](https://img.shields.io/static/v1?style=default&label=Discord&logo=discord&message=join)](https://discord.gg/xC7ucM8j)
[![CI status](https://github.com/learning-at-home/hivemind/actions/workflows/run-tests.yml/badge.svg?branch=master)](https://github.com/learning-at-home/hivemind/actions)
![Codecov](https://img.shields.io/codecov/c/github/learning-at-home/hivemind)
[![Black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

Hivemind is a PyTorch library for decentralized deep learning across the Internet. Its intended usage is training one
large model on hundreds of computers from different universities, companies, and volunteers.

![img](https://i.imgur.com/GPxolxb.gif)

## Key Features

* Distributed training without a master node: Distributed Hash Table allows connecting computers in a decentralized
  network.
* Fault-tolerant backpropagation: forward and backward passes succeed even if some nodes are unresponsive or take too
  long to respond.

> * [Decentralized parameter averaging](https://arxiv.org/abs/2103.03239): iteratively aggregate updates from multiple
>   workers without the need to synchronize across the entire network.

Reviewer: ```suggestion
* Decentralized parameter averaging: iteratively aggregate updates from multiple
  workers without the need to synchronize across the entire network ([paper](https://arxiv.org/abs/2103.03239)).
```

This way, a reader will see that we refer to our paper (not a Wikipedia definition), which may be interesting to open for them.

* Train neural networks of arbitrary size: parts of their layers are distributed across the participants with the

>   [Decentralized Mixture-of-Experts](https://arxiv.org/abs/2002.04013).

Reviewer: ```suggestion
  decentralized mixture-of-experts ([paper](https://arxiv.org/abs/2002.04013)).
```

## Hivemind: decentralized deep learning in PyTorch

[![Documentation Status](https://readthedocs.org/projects/learning-at-home/badge/?version=latest)](https://learning-at-home.readthedocs.io/en/latest/?badge=latest)
[![PyPI version](https://img.shields.io/pypi/v/hivemind.svg)](https://pypi.org/project/hivemind/)

> [![Discord](https://img.shields.io/static/v1?style=default&label=chat&logo=discord&message=join%20chat)](https://discord.gg/xC7ucM8j)

Reviewer 2: ```suggestion
[![Discord](https://img.shields.io/static/v1?style=default&label=Discord&logo=discord&message=join)](https://discord.gg/xC7ucM8j)
```

[![CI status](https://github.com/learning-at-home/hivemind/actions/workflows/run-tests.yml/badge.svg?branch=master)](https://github.com/learning-at-home/hivemind/actions)
![Codecov](https://img.shields.io/codecov/c/github/learning-at-home/hivemind)
[![Black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

Hivemind is a PyTorch library for decentralized deep learning across the Internet. Its intended usage is training one
large model on hundreds of computers from different universities, companies, and volunteers.

![img](https://i.imgur.com/GPxolxb.gif)

## Key Features

* Distributed training without a master node: Distributed Hash Table allows connecting computers in a decentralized
  network.
* Fault-tolerant backpropagation: forward and backward passes succeed even if some nodes are unresponsive or take too
  long to respond.
* [Decentralized parameter averaging](https://arxiv.org/abs/2103.03239): iteratively aggregate updates from multiple
  workers without the need to synchronize across the entire network.

> * Train neural networks of arbitrary size: parts of their layers are distributed across the participants with

Reviewer 2: ```suggestion
* Train neural networks of arbitrary size: parts of their layers are distributed across the participants with the
```

## Key Features

> * Train neural networks of arbitrary size: parts of their layers are distributed across the participants.

Reviewer 3: We can describe MoE somewhere in readme, but this sentence here was totally misleading

> * Train neural networks of arbitrary size: parts of their layers are distributed across the participants.

Reviewer 3: We can describe MoE somewhere in readme, but this sentence here was totally misleading

If you would like to verify that your installation is working properly, you can install with `pip install -e .[dev]`
instead. Then, you can run the tests with `pytest tests/`.

By default, hivemind uses the precompiled binary of
the [go-libp2p-daemon](https://github.com/learning-at-home/go-libp2p-daemon) library. If you face compatibility issues
or want to build the binary yourself, you can recompile it by running `pip install . --global-option="--buildgo"`.
Before running the compilation, please ensure that your machine has a recent version
of [Go toolchain](https://golang.org/doc/install) (1.15 or higher).

> ### OS support

Reviewer: ```suggestion
### System requirements
```

- __Linux__ is the default OS for which hivemind is developed and tested. We recommend Ubuntu 18.04+ (64-bit),
  but other 64-bit distros should work as well. Legacy 32-bit is not recommended.
- __macOS 10.x__ mostly works but requires building hivemind from source, and some edge cases may fail.
  To ensure full compatibility, we recommend using [our Docker image](https://hub.docker.com/r/learningathome/hivemind).

> - __Windows 10+ (experimental)__ hivemind can be run using [WSL](https://docs.microsoft.com/ru-ru/windows/wsl/install-win10).

Reviewer: ```suggestion
- __Windows 10+ (experimental)__ can run hivemind using [WSL](https://docs.microsoft.com/ru-ru/windows/wsl/install-win10).
```

  You can configure WSL to use GPU following [this guide](https://docs.nvidia.com/cuda/wsl-user-guide/index.html) by NVIDIA.

>   After CUDA toolkit is you should install hivemind with pip or build it from source using the instructions above.

Reviewer 2: ```suggestion
  After the CUDA toolkit is installed, you can install hivemind with pip or build it from source using the instructions above.
```

## Documentation

> * [Quickstart tutorial](https://learning-at-home.readthedocs.io/en/latest/user/quickstart.html) walks through installation

Reviewer: ```suggestion
* The [quickstart tutorial](https://learning-at-home.readthedocs.io/en/latest/user/quickstart.html) walks through installation
```

>   and a training a simple neural network with multiple  

Reviewer 2: ?

Author: [fixd]

* [examples/albert](https://github.com/learning-at-home/hivemind/tree/master/examples/albert) contains the starter kit
  and instructions for training a Transformer masked language model collaboratively.
* API reference and additional tutorials are available at [learning-at-home.readthedocs.io](https://learning-at-home.readthedocs.io)

> If you have any questions about intalling and using hivemind, you can ask them in 

Reviewer 2: ```suggestion
If you have any questions about installing and using hivemind, you can ask them in 
```

> [our discord chat](https://discord.gg/xC7ucM8j) or file an [issue](https://github.com/learning-at-home/hivemind/issues).

Reviewer 2: ```suggestion
[our Discord chat](https://discord.gg/xC7ucM8j) or file an [issue](https://github.com/learning-at-home/hivemind/issues).
```

}
```

<details>
 <summary>Additional publications</summary>

> ["Moshpit SGD: Communication-Efficient Decentralized Training on Heterogeneous Unreliable Devices"](https://arxiv.org/abs/2103.03239)

Reviewer: Maybe specify the last names of all author or the first authors to highlight people contributed to the project?
